File: problems/0048-rotate-image/solution.go
func rotate(matrix [][]int)  {
    // 1. flip across top-left to bottom-right axis 
    // 2. reverse each row
    n := len(matrix)
    m := len(matrix[0])


    for i := 0; i < n; i++ {
        for j := 0; j < i; j++ {
            matrix[i][j], matrix[j][i] = 
              matrix[j][i], matrix[i][j]
        }
    }

   
    for i := 0; i < n; i++ {
        for j := 0; j < m / 2; j++ {
            matrix[i][j], matrix[i][m-j-1] = matrix[i][m-j-1], matrix[i][j]
        }
    }
}

File: problems/0048-rotate-image/solution.cpp
class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        // 1. flip along top left to bottom right axis
        // 2. reverse each row


        int n = matrix.size();
        int m = matrix[0].size();

        for (int i = 0; i < n; i++) {
            for (int j = i; j < m; j++) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = tmp;
            }
        }

        for (int i = 0; i < n; i++) {
            // reverse the row
            for (int j = 0; j < int(n / 2); j++) {
                int tmp = matrix[i][j];
                matrix[i][j] = matrix[i][m-j-1];
                matrix[i][m-j-1] = tmp;
            }
        }
        
    }
};
